### ISSUE TITLE
eos_token not being encoded the same after finetuning and hence not being detected at inference time

### ISSUE THREAD

**Issue by iliasmiraoui** [huggingface/peft#1003]
### System Info

Using peft 0.5.0 & transformers 4.35.0.dev0.

I just finetuned mistralai/Mistral-7B-Instruct-v0.1 on some instruction-data and it seems like the eos_token is not the same anymore on the fine-tuned version of the model. 

On the base model, `</s>` gets encoded as "2", regardless of what comes before it (Even when it is "}</s>"). On the LoRA I finetuned however, It gets encoded into four different tokens with "}</" getting encoded as "10050".

My instruction fine-tuning data looks like this:
`<s>[INST] Instruction 1 [/INST] {'some_key':'answer 1'}</s>`

This then creates a ton of issues with model.generate as the model doesn't stop as it should, etc. Am I doing something wrong or is LoRA changing the eos_token?

### Who can help?

@pacman100 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

My instruction fine-tuning data looks like this:
`<s>[INST] Instruction 1 [/INST] {'some_key':'answer 1'}</s>`

### Expected behavior

On the base model, `</s>` gets encoded as "2", regardless of what comes before it (Even when it is "}</s>"). On the LoRA I finetuned however, It gets encoded into four different tokens with "}</" getting encoded as "10050". I would expcted this to be encoded properly into 2 so that model.generate could pick it up correctly as the eos_token_id

**Comment by iliasmiraoui** [comment]
<img width="732" alt="Screenshot 2023-10-06 at 5 11 00 PM" src="https://github.com/huggingface/peft/assets/83364633/4a697c5a-817d-4092-bcf0-7ecb55a781c4">

Example of how "2" gets normally encoded by the tokenizer but how "10050", which is actually being output by the model

**Comment by BenjaminBossan** [comment]
Thanks for reporting. Could you please provide more context, especially the code how you fine-tune the model with PEFT.

The issue you describe seems very strange to me, as using LoRA should have no influence on the tokenizer, so I think I'm missing something here. Do you mean that the output of your last screenshot is different after the model was fine-tuned with PEFT? What was it before?

**Comment by iliasmiraoui** [comment]
Here is the code I used to train the model:
`
model_name = "mistralai/Mistral-7B-Instruct-v0.1"
################################################################################
# QLoRA parameters
################################################################################

# LoRA attention dimension
lora_r = 64

# Alpha parameter for LoRA scaling
lora_alpha = 16

# Dropout probability for LoRA layers
lora_dropout = 0.1

################################################################################
# bitsandbytes parameters
################################################################################

# Activate 4-bit precision base model loading
use_4bit = True

# Compute dtype for 4-bit base models
bnb_4bit_compute_dtype = "float16"

# Quantization type (fp4 or nf4)
bnb_4bit_quant_type = "nf4"

# Activate nested quantization for 4-bit base models (double quantization)
use_nested_quant = False

################################################################################
# TrainingArguments parameters
################################################################################

# Output directory where the model predictions and checkpoints will be stored
output_dir = "./results"

# Number of training epochs
num_train_epochs = 2

# Enable fp16/bf16 training (set bf16 to True with an A100)
fp16 = False
bf16 = True

# Batch size per GPU for training
per_device_train_batch_size = 4

# Batch size per GPU for evaluation
per_device_eval_batch_size = 4

# Number of update steps to accumulate the gradients for
gradient_accumulation_steps = 1

# Enable gradient checkpointing
gradient_checkpointing = True

# Maximum gradient normal (gradient clipping)
max_grad_norm = 0.3

# Initial learning rate (AdamW optimizer)
learning_rate = 2e-5

# Weight decay to apply to all layers except bias/LayerNorm weights
weight_decay = 0.001

# Optimizer to use
optim = "paged_adamw_32bit"

# Learning rate schedule
lr_scheduler_type = "cosine"

# Number of training steps (overrides num_train_epochs)
max_steps = -1

# Ratio of steps for a linear warmup (from 0 to learning rate)
warmup_ratio = 0.03

# Group sequences into batches with same length
# Saves memory and speeds up training considerably
group_by_length = True

# Save checkpoint every X updates steps
save_steps = 0

# Log every X updates steps
logging_steps = 50

################################################################################
# SFT parameters
################################################################################

# Maximum sequence length to use
max_seq_length = None

# Pack multiple short examples in the same input sequence to increase efficiency
packing = False

# Load the entire model on the GPU 0
device_map = {"": 0}

# Load tokenizer and model with QLoRA configuration
compute_dtype = getattr(torch, bnb_4bit_compute_dtype)

bnb_config = BitsAndBytesConfig(
    load_in_4bit=use_4bit,
    bnb_4bit_quant_type=bnb_4bit_quant_type,
    bnb_4bit_compute_dtype=compute_dtype,
    bnb_4bit_use_double_quant=use_nested_quant,
)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map=device_map,
    quantization_config=bnb_config,
    #use_flash_attention_2=True
)

model.config.use_cache = False
model.config.pretraining_tp = 1

# Load LLaMA tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
tokenizer.pad_token = tokenizer.eos_token

peft_config = LoraConfig(
    lora_alpha=lora_alpha,
    lora_dropout=lora_dropout,
    r=lora_r,
    inference_mode=False,
    task_type="CAUSAL_LM",
    target_modules = ["q_proj", "v_proj"]
)

# Set training parameters
training_arguments = TrainingArguments(
    output_dir=output_dir,
    num_train_epochs=num_train_epochs,
    per_device_train_batch_size=per_device_train_batch_size,
    gradient_accumulation_steps=gradient_accumulation_steps,
    optim=optim,
    save_steps=save_steps,
    logging_steps=logging_steps,
    eval_steps=logging_steps,
    evaluation_strategy="steps",
    learning_rate=learning_rate,
    weight_decay=weight_decay,
    fp16=fp16,
    bf16=bf16,
    max_grad_norm=max_grad_norm,
    max_steps=max_steps,
    warmup_ratio=warmup_ratio,
    group_by_length=group_by_length,
    lr_scheduler_type=lr_scheduler_type,
    report_to="tensorboard"
)

# Set supervised fine-tuning parameters
trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    eval_dataset=eval_dataset,
    peft_config=peft_config,
    dataset_text_field="text",
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=packing,
)
trainer.train()`


I used the screenshots to show that the token decode to the same thing after the finetune than the base model, but instead of the eos token being encoded as 2, the model outputs 10050/etc.

**Comment by BenjaminBossan** [comment]
So if I understand correctly, you're saying that the fine-tuned model predicts `}</s>` as `[10050, 28713, 28767]` but you would expect `[..., 2]`, is that right? Does the same issue occur when you don't pass `peft_config` to `SFTTrainer`?

**Comment by iliasmiraoui** [comment]
that's right !  I haven't been able to try yet because I would need more GPU to do a full finetune. I will try to give it a shot

**Comment by BenjaminBossan** [comment]
That would be great, as I suspect that this is not due to PEFT. Maybe you can try a quick training run with just a small amount data, as the problem probably does not require a full training run to surface.

**Comment by iliasmiraoui** [comment]
Will do

**Comment by iliasmiraoui** [comment]
Was able to narrow down the cause and a related problem on the transformers lib: https://github.com/huggingface/transformers/issues/26743

The tokenizer tokenizes the eos_token differently depending on context for mistral so it requires extra-care.

**Comment by hengjiUSTC** [comment]
@iliasmiraoui hey dude, how to fix then? i have a finetuned model (very similar with your training process) loaded by:
```
model_name = "NousResearch/Llama-2-7b-chat-hf"
device_map = {"": 0}
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch.float16,
    device_map=device_map,
)
model = PeftModel.from_pretrained(base_model, "HenryJJ/tangshi-llama2-7b-chat-qlora")
```
and inference with:
```
input_ids = tokenizer(prompt, return_tensors="pt").input_ids.cuda()
# with torch.inference_mode():
outputs = model.generate(
    input_ids=input_ids,
    max_new_tokens=400,
    top_p=0.9,
    do_sample=True,
    temperature=0.7,
    eos_token_id=tokenizer.eos_token_id
)
```
I keep seeing `</s>` on my result :(
